service/controller/v14/resource/bridgezone/resource.go

```diff
@@ -0,0 +1,415 @@
+		}
+		_, err := defaultGuest.ChangeResourceRecordSetsWithContext(ctx, in)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", "ensured intermediate zone delegation")
+	}
+
+	return nil
+}
+
+func (r *Resource) EnsureDeleted(ctx context.Context, obj interface{}) error {
+	customObject, err := key.ToCustomObject(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	if !r.route53Enabled {
+		r.logger.LogCtx(ctx, "level", "debug", "message", "route53 disabled, skipping execution")
+		return nil
+	}
+
+	baseDomain := key.BaseDomain(customObject)
+	intermediateZone := "k8s." + baseDomain
+	finalZone := key.ClusterID(customObject) + ".k8s." + baseDomain
+
+	_, defaultGuest, err := r.route53Clients(ctx)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	var intermediateZoneID string
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", "searching for intermediate zone in default guest account")
+
+		intermediateZoneID, err = r.findHostedZoneID(ctx, defaultGuest, intermediateZone)
+		if IsNotFound(err) {
+			r.logger.LogCtx(ctx, "level", "debug", "message", "intermediate zone does not exist, skipping the resource")
+			return nil
+		} else if err != nil {
+			return microerror.Mask(err)
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", "intermediate zone found")
+	}
+
+	var finalZoneTTL int64
+	var finalZoneRecords []*route53.ResourceRecord
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", "getting intermediate zone name servers")
+
+		nameServers, ttl, err := r.getNameServersAndTTL(ctx, defaultGuest, intermediateZoneID, finalZone)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		finalZoneTTL = ttl
+
+		for _, ns := range nameServers {
+			copy := ns
+			v := &route53.ResourceRecord{
+				Value: &copy,
+			}
+			finalZoneRecords = append(finalZoneRecords, v)
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", "got intermediate zone name servers")
+	}
+
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", "ensuring deletion of intermediate zone delegation")
+
+		delete := route53.ChangeActionDelete
+		ns := route53.RRTypeNs
+
+		in := &route53.ChangeResourceRecordSetsInput{
+			ChangeBatch: &route53.ChangeBatch{
+				Changes: []*route53.Change{
+					{
+						Action: &delete,
+						ResourceRecordSet: &route53.ResourceRecordSet{
+							Name:            &finalZone,
+							Type:            &ns,
+							TTL:             &finalZoneTTL,
+							ResourceRecords: finalZoneRecords,
+						},
+					},
+				},
+			},
+			HostedZoneId: &intermediateZoneID,
+		}
+		_, err := defaultGuest.ChangeResourceRecordSetsWithContext(ctx, in)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", "deletion of intermediate zone delegation ensured")
+	}
+
+	return nil
+}
+
+func (r *Resource) findHostedZoneID(ctx context.Context, client *route53.Route53, name string) (string, error) {
```

Comment:
Would it make sense to have unit tests for this?

Reply:
W/o mocking the whole client and making potential wrong assumptions this is the only part that is testable:

```
+			if hz.Name == nil || hz.Id == nil {
+				continue
+			}
+
+			hzName := *hz.Name
+			hzName = strings.TrimSuffix(hzName, ".")
+			hzID := *hz.Id
+
+			if hzName == name {
+				return hzID, nil
+			}
```

This is 3 LoC. I'm pretty confident it works as expected.